### ISSUE TITLE
Extend the usage of "Template for new translations:" and according Translation States

### ISSUE THREAD

**Issue by rhofer** [WeblateOrg/weblate#3273]
## Situation

The current component configuration allows to specify:
- **Filemask:**
- **Monolingual base language file:**
- [ ] **Edit base file**
- **Template for new translations:**

As per my knowledge *Template* is only used when creating a new translation (language) to create all the keys available from *Template* but with empty strings. This could be extended to match the requirements for a quality gate described in #3270 

## Solution Approach

Assuming the *Template* provided also includes strings, what are considered as SWEnglish or Developer-English. This, in a component with a configuration such as (e.g. json use case):
- **Filemask:** `src/assets/i18n/*.json`
- **Monolingual base language file:** `src/assets/i18n/en.json`
- [x] **Edit base file** (allowed)
- **Template for new translations:** `src/assets/i18n/template.json`

Then the quality gate related usecases may shall work as following:

### Usecase 1: New component, no `en`, yet

- Configure a new component, where currently only `template.json` as file is available
- Weblate starts importing the connected VCS, detects that base file shall be `en.json` what is not yet there.
- Weblate creates the `en.json` based on `template.json` , including keys AND provided strings (SWEnglish), but marks all strings in `en` with `needs editing`
- Since `en` as base language is editable, the harmonization / polishing work can directly happen on this without the need to artificially create a language called SWEnglish/template

### Usecase 2: Update of *Template* and/or `en` already exists

Assuming `en` already has been created, e.g. according usecase 1.

1. Developer **adds** new key-value-pair to *Template*:
- key-value-pair shall be added to `en` and marked with *needs editing*, as per usecase 1

2. Developer **changes** value of existing key in *Template*:
- Somehow on `en` a kind of notification shall happen, that the SWEnglish has changed
  - Override the existing string in `en` with the one from *Template* and mark as needs editing. The *History* tab would then reveal , that this is an update from *Template* where previously already a harmonized (and approved) string has been. Since it is marked now as *needs editing*, the string coming newly from *Template* would then not be automatically propagated to further translations. See usecase 4.
  - ... any other solution to notify the owners of `en` that SWEnglish coming from *Template* has changed?

3. Developer **deletes** key-value-pair in *Template*:
- Since the context provisionning (keys) is based on *Template*, the offering of this string for translation in `en` and any further languages shall be hidden. This is as it already works today, but instead of referring to *Monolingual base language file:* (source) it shall refer to *Template*
- If the string already has been added once to `en` and any further languages, it shall remain in the related VCS' files (`en.json`, `fr.json`, etc.), unless the appropriate Addon for this component is active as *Cleanup translation files*

### Usecase 3: Create another language for translation

For this, the existing mechanisms shall remain, means, to take the keys from *Template* in order to create this language. But of course, the values (SWEnglish) shall not be copied to other languages than `en`

### Usecase 4: Translate in further language based on `en`

Further translations shall depend on Translation states in `en`
- If a string in `en` is marked with *needs editing*, the translation work in a further language for this string only is "blocked"
  - Don't show the string at all
  - Disable Save button
  - ... whatever is the best solution
- If this string goes to *waiting for review* or even *approved*, it shall become translatable in another language

### Usecase 5: *Template* has empty string

This depends on the targeted workflow and may requires an additional checkbox in the component configuration besides the *Template for new translations:* input field, such as:
- **Template for new translations:** `src/assets/i18n/template.json`
- [x] Template acts as source for developer-terms / developer-english (true/false)

This may leads to:
- If Template acts as source is true, all the above usecases shall apply
- If its false, it shall behave as today (only key provider for new translations)

### Translation Memory

- If any string of any language is in state *needs editing* it shall not go into TM (as it works today)

### Machine Translation

- Machine translation provides results from TM and from *Components* (either itself or other components)
- If another *Component* would have (partially) matching strings but it's source string (e.g. from `en`) is marked as *needs editing*, the machine shall NOT show/offer this result.
- If a string from another *Component* once went from *needs editing* to *approved* and now back to *needs editing*, the originally *approved* string anyways ended up in TM, hence this result (originally approved one) shall be provided / offered as TM result.

### Addons

Impact on addons must be evaluated, overall. What is currently ovious to me:

#### Addon: Cleanup translation files

- In the usecases above, this addon shall work against *Template for new translations* and not against *Monolingual base language file*
- Possibly to be solved with the addon itself to make it configurable in order to not strictly match the monolingual base file but with options:
  1. monolingual base file
  2. template

**Comment by nijel** [comment]
I still don't get whether having separate template language is needed or it's just a way that work. I see no reason why this all could not happen directly on the English file. All new/changed strings would be marked for review (blocking translations until review happens). With two separate files, things just get more complex (you would have to deal with situations where key is changed on both sides).

**Comment by rhofer** [comment]
We designed the workflow intentionally with a template file (it is not necessarely a language in a sense of a natural language). This allows that the SW-Project
-  is the owner of this template file.
- only they edit this template file
- weblate only consumes / never modifies

On the other way round, weblate
- owns all translation results incl. `en`
- The software project never modifies these results, only consumes 

With this we installed a one directional flow in each direction, hence we never have merge conflicts. Furthermore, there never is a conflicting key change over multiple files. Keys only change in this template file.

Secondly, in some cases this template file is a build artifact and not checked in to the repos. Hence, modifying this template file in weblate makes no sense, since with every build it will be overwritten. So from this point of view to replace template with en directly would not be possible, since we follow the paradigm that build artifacts are never checked in to the same repository.

I'm not sure, whether this makes it clearer to understand. Possibly I also missed something, what would make my approach easier ...
